### PR TITLE
Feature/2.x/261 open referral

### PR DESCRIPTION
**Pull request by ekes**
- [ ] Requires localgovdrupal/localgov_openreferral#14
 - [ ] Also #138

### CODE REVIEW

File: modules/localgov_directories_or/localgov_directories_or.module
Comment (Adnan-cds) [inline modules/localgov_directories_or/localgov_directories_or.module]: s/syncronise/synchronise/ ?

File: modules/localgov_directories_or/src/FacetMapping.php
Comment (Adnan-cds) [inline modules/localgov_directories_or/src/FacetMapping.php]: s/\Drupal::entityTypeManager()->/$this->entityTypeManager->/ ?

File: modules/localgov_directories_org/config/install/node.type.localgov_directories_org.yml
Comment (Adnan-cds) [inline modules/localgov_directories_org/config/install/node.type.localgov_directories_org.yml]: s/Either or/Either as/ ?

File: modules/localgov_directories_or/src/FacetMapping.php
Comment (Adnan-cds) [inline modules/localgov_directories_or/src/FacetMapping.php]: Err..

File: modules/localgov_directories_or/src/FacetMapping.php
Comment (Adnan-cds) [inline modules/localgov_directories_or/src/FacetMapping.php:100]: I have a feeling FacetMapping::syncroniseFacetMappings() has two methods worth of code within it.  This leaves some scope for improving code readability and thus DX.

**Review by Adnan-cds — COMMENTED** [review]
RIP (Review In Progress)

File: modules/localgov_directories_or/localgov_directories_or.info.yml
Comment (Adnan-cds) [inline modules/localgov_directories_or/localgov_directories_or.info.yml:9]: localgovdrupal/localgov_openreferral is not mentioned in localgov_directories' composer.json.  I don't know if this is right or wrong. I have had this issue in another [pull request](https://github.com/localgovdrupal/localgov_alert_banner/blob/3ad05ca8fce1d81fe937df771be428ece19ee190/composer.json#L10) where I hesitated to add submodule dependencies in the parent module's  composer.json file. @andybroomfield @ekes @finnlewis @stephen-cox 

Please note that the localgovdrupal/localgov_openreferral package doesn't exist in packagist.org at this time.

**Comment by Adnan-cds** [comment]
The following config file adds a menu entry for **Directory organization** into the *Content > Add content > Directory* menu link group:
```
langcode: en
status: true
dependencies:
  enforced:
    module:
      - localgov_directories
      - localgov_menu_link_group
      - localgov_directories_org
id: localgov_directories_org
group_label: Directory
weight: 0
parent_menu: admin
parent_menu_link: 'admin_toolbar_tools.extra_links:node.add'
child_menu_links:
  - 'admin_toolbar_tools.extra_links:node.add.localgov_directories_org'
```
Location: ```modules/localgov_directories_org/config/optional/localgov_menu_link_group.localgov_menu_link_group.localgov_directories_org.yml```

**Review by Adnan-cds — COMMENTED** [review]
The core functionality seems to be working.  I added a new directory channel, a new facet type called **location**, and a few facet items.  When I then hit /openreferral/v1/taxonomies?vocabulary=location, all the facet items were returned in JSON format.

I also enabled the localgov_directories_org module and its title and body content were also returned from /openreferral/v1/organiztion/ORG-NODES-UUID

Few things to notes:
- It will be good to have a README file.
- /openreferral/v1/taxonomies is throwing a WSOD instead of a 404.  I think the [relevant code](https://github.com/localgovdrupal/localgov_openreferral/blob/d8062bbc32922a20999e3e12966f32d0ace43ac8/src/Controller/EndpointsController.php#L107) needs to use a different exception class.
- localgov_directories_org submodule has a missing dependency on the localgov_directories_venue submodule.

I used the following **root level** composer.json file in my localgov test site for this branch:
```
{
    "name": "localgovdrupal/localgov-project",
    "description": "Project template for the LocalGov Drupal distribution.",
    "type": "project",
    "license": "GPL-2.0-or-later",
    "homepage": "https://github.com/localgovdrupal/localgov_project",
    "repositories": [
        {
            "type": "vcs",
            "url": "https://git.drupalcode.org/issue/geocoder_ajax_prepopulate-3182650.git"
        },
        {
            "type": "composer",
            "url": "https://packages.drupal.org/8"
        }
    ],
    "require": {
        "composer/installers": "^1.10",
        "cweagans/composer-patches": "^1.6",
        "drupal/core-composer-scaffold": "^9.1",
        "drupal/core-project-message": "^9.1",
        "drupal/core-recommended": "^9.1",
        "localgovdrupal/localgov": "^2.0",
        "localgovdrupal/localgov_directories": "dev-feature/2.x/261-Open-Referral as 2.0.2",
        "localgovdrupal/localgov_geo": "dev-feature/1.x/area-and-ux as 1.x-dev"
    },
    "require-dev": {
        "brianium/paratest": "^6.0",
        "drupal/coder": "^8.3",
        "drupal/config_inspector": "^1.3",
        "drupal/core-dev": "^9.1",
        "mglaman/phpstan-drupal": "0.12.10",
        "phpspec/prophecy-phpunit": "^2",
        "phpstan/phpstan": "^0.12",
        "phpstan/phpstan-deprecation-rules": "^0.12",
        "squizlabs/php_codesniffer": "^3.6"
    },
    "conflict": {
        "drupal/drupal": "*"
    },
    "minimum-stability": "dev",
    "prefer-stable": true,
    "config": {
        "sort-packages": true,
        "bin-dir": "bin/"
    },
    "extra": {
        "enable-patching": true,
        "drupal-scaffold": {
            "locations": {
                "web-root": "web/"
            },
            "file-mapping": {
                "[web-root]/sites/default/settings.php": {
                    "mode": "replace",
                    "path": "assets/composer/settings.php",
                    "overwrite": false
                },
                "[web-root]/sites/default/settings.lando.php": {
                    "mode": "replace",
                    "path": "assets/composer/settings.lando.php",
                    "overwrite": false
                },
                "[web-root]/sites/development.services.yml": "assets/composer/development.services.yml"
            }
        },
        "installer-paths": {
            "web/core": ["type:drupal-core"],
            "web/libraries/{$name}": ["type:drupal-library"],
            "web/modules/contrib/{$name}": ["type:drupal-module"],
            "web/profiles/contrib/{$name}": ["type:drupal-profile"],
            "web/themes/contrib/{$name}": ["type:drupal-theme"],
            "drush/Commands/contrib/{$name}": ["type:drupal-drush"],
            "web/modules/custom/{$name}": ["type:drupal-custom-module"],
            "web/themes/custom/{$name}": ["type:drupal-custom-theme"]
        },
        "drupal-core-project-message": {
            "post-create-project-cmd-message": [
                "<bg=blue;fg=white>                                                                  </>",
                "<bg=blue;fg=white>  Congratulations, you’ve installed the LocalGov Drupal codebase  </>",
                "<bg=blue;fg=white>                                                                  </>",
                "",
                "<bg=yellow;fg=black>Next steps</>:",

                "  * Install the site: https://www.drupal.org/docs/8/install",
                "  * Remove the plugin that prints this message:",
                "      composer remove drupal/core-project-message"
            ]
        }

    }
}
```
Because the localgov_openreferral module is not available as a composer package, I just had to git clone it.

This is the **end** of my review.  I am neither approving it nor disapproving it because not all tests are passing.  Also, I have pointed out a few small issue which you may want to fix.  In any case, thanks a lot.

File: modules/localgov_directories_or/localgov_directories_or.info.yml
Comment (andybroomfield) [reply modules/localgov_directories_or/localgov_directories_or.info.yml:9]: On leave this week... however...

My suggestion was in cases where it is a sub module to use composer suggest. Though in this case if localgov_directories_org is for open referral, and that's where the functionality is, then maybe this module should be a submodule of localgov_openreferral instead and that should have a dependency on localgov_directories.

Comment (ekes) [reply modules/localgov_directories_or/localgov_directories_or.info.yml:9]: Think most of the comments above make sense, shall get to them when looking at this again next week.

On this particular one however:-
 * LocalGov Open Referral will in time have a packagist entry etc. Hadn't got that far yet.
 * LocalGov Directories Organisation is just a Directory type. It can be used as is without Open Referral should you want. It's not even required if you want Directories to output Open Referral. Something mapped to Open Referral Organisation type is required and this module offers it to you without having to do anything custom.
 * LocalGov Open Referral has no code about Directories at all. In theory you can use it with other content types.

So a suggest in the LocalGov Directories composer.json makes sense I think.

File: modules/localgov_directories_or/src/FacetMapping.php
Comment (ekes) [reply modules/localgov_directories_or/src/FacetMapping.php:100]: It's two loops, the first making a list used by the second. They both work with the same entity storage for much the same purposes. It seems a bit much to need to split apart and call from the same function at this point?

File: modules/localgov_directories_or/localgov_directories_or.info.yml
Comment (ekes) [reply modules/localgov_directories_or/localgov_directories_or.info.yml:9]: Gone with the suggests.

**Comment by ekes** [comment]
> It will be good to have a README file.

Yes I think some documentation about enabling OpenReferral in Directories would be good. Added #141 - not sure if it needs to block this?

**Comment by Adnan-cds** [comment]
> localgov_directories_org submodule has a missing dependency on the localgov_directories_venue submodule.

Thanks a lot Ekes for all the changes from yesterday.  The above issue is still happening :(  Here's the error message when I tried to install the localgov_directories_org module on a **fresh install** of Localgov:

> In UnmetDependenciesException.php line 100:
> 
>  Configuration objects provided by <em class="placeholder">localgov_directories_org</em> ha
>  ve unmet dependencies: <em class="placeholder">field.field.node.localgov_directories_org.l
>  ocalgov_directory_notes (field.storage.node.localgov_directory_notes), field.field.node.lo
>  calgov_directories_org.localgov_location (localgov_geo.localgov_geo_type.address)</em>

Looking good otherwise.  Nearly there.

**Comment by ekes** [comment]
> field.storage.node.localgov_directory_notes

Where did that get added? I'll look at that.

> field.field.node.localgov_directories_org.localgov_location (localgov_geo.localgov_geo_type.address

This PR depends not only on having localgov_openreferral; but also the localgov_geo with address module (current version of the PR https://github.com/localgovdrupal/localgov_geo/pull/29 - which is well worth doing a review on ;-)

So for now I'd say order of dependencies is:-
 * Get the Localgov Geo PR in
 * Change the Localgov Directories Venue dependency from localgov_geo to localgov_geo_address
 * Get Localgov OpenReferral into the default code base (but not default enabled)
 * Get this PR in (it will then have both the localgov_geo_address, with label, dependency; and the OpenReferral module dependency).

**Comment by Adnan-cds** [comment]
> but also the localgov_geo with address module (current version of the PR localgovdrupal/localgov_geo#29

I have been using *localgovdrupal/localgov_geo:dev-feature/1.x/area-and-ux* in the test site and it does have the localgov_geo_address submodule.  I have just switched to *localgovdrupal/localgov_geo:dev-feature/1.x/area-and-ux-autocomplete* but the UnmetDependenciesException persists :(  I am assuming the "*Change the Localgov Directories Venue dependency from localgov_geo to localgov_geo_address*" action you have mentioned above will solve this once you implement it.

**Comment by ekes** [comment]
> I am assuming the "Change the Localgov Directories Venue dependency from localgov_geo to localgov_geo_address" action you have mentioned above will solve this once you implement it.

So #143 first. Then this.
And before #143 localgovdrupal/localgov_geo#29

**Comment by Adnan-cds** [comment]
> So #143 first. Then this.

Okay, l will wait until #143 is merged.

**Comment by Adnan-cds** [comment]
I am reviewing again now that all the dependencies are in place.

**Comment by Adnan-cds** [comment]
Ekes, just to be sure, does the 2.x branch need to be merged with this branch first?

**Comment by Adnan-cds** [comment]
Hi Ekes, so I merged 2.x with feature/2.x/261-Open-Referral in a fresh localgov install in my local machine but still having issues :(

```
$ drush -y en localgov_directories_venue
The following module(s) will be enabled: localgov_directories_venue, localgov_directories, search_api, search_api_db, facets, localgov_directories_location, localgov_geo, geofield, geofield_map, leaflet, leaflet_views

 // Do you want to continue?: yes.
*Usual Search API error that we just ignore goes here - adnan*

In UnmetDependenciesException.php line 100:
 
  Configuration objects provided by <em class="placeholder">localgov_directories_venue</em>
  have unmet dependencies: <em class="placeholder">field.field.node.localgov_directories_ven
  ue.localgov_location (localgov_geo.localgov_geo_type.address)</em>
```

And then
```
$ drush -y en localgov_directories_org

In UnmetDependenciesException.php line 100:
 
  Configuration objects provided by <em class="placeholder">localgov_directories_org</em> ha
  ve unmet dependencies: <em class="placeholder">field.field.node.localgov_directories_org.l
  ocalgov_directory_notes (field.storage.node.localgov_directory_notes), field.field.node.lo
  calgov_directories_org.localgov_location (localgov_geo.localgov_geo_type.address)</em>
```

If you can please take a look.

There were also error messages **before** I merged 2.x with feature/2.x/261-Open-Referral but let's ignore those.

**Comment by Adnan-cds** [comment]
The "UnmetDependenciesException.php" exceptions are gone.  Thanks Ekes :)

**Comment by Adnan-cds** [comment]
> I added a new directory channel, a new facet type called location, and a few facet items. When I then hit /openreferral/v1/taxonomies?vocabulary=location, all the facet items were returned in JSON format.

This one isn't working any more in my test site.  Not sure if I have missed anything.  Could you please comfirm Ekes if this works for you.

> I also enabled the localgov_directories_org module and its title and body content were also returned from /openreferral/v1/organiztion/ORG-NODES-UUID

This works fine :)

File: modules/localgov_directories_venue/localgov_directories_venue_or/localgov_directories_venue_or.info.yml
Comment (stephen-cox) [inline modules/localgov_directories_venue/localgov_directories_venue_or/localgov_directories_venue_or.info.yml]: Perhaps move this module out of the `localgov_directories_venue` directory so it's in the `modules` dir so it's easier to find and more consistent.

File: composer.json
Comment (stephen-cox) [inline composer.json]: De want to set this to a particular version? E.g. `1.x-dev`?

**Comment by andybroomfield** [comment]
When testing this branch with demo content I'm seeing this error

> The website encountered an unexpected error. Please try again later.
> Error: Call to a member function getCacheTags() on null in Drupal\localgov_directories\Plugin\Block\ChannelSearchBlock->getCacheTags() (line 144 of modules/contrib/localgov_directories/src/Plugin/Block/ChannelSearchBlock.php)

When viewing citizen advice entry.

**Comment by ekes** [comment]
Notes about paths urls from Merge Monday review chat:

https://localgov.lndo.site/openreferral/v1/services -- that's the view (list) requires search api index to have run ;-)

The rest are in the localgov_openreferral.routing.yml so eg https://localgov.lndo.site/openreferral/v1/organization/d2b6dc2a-848e-4d86-aa4d-020d8bbd5b83 would be the organization node with the uuid d2b6dc2a-848e-4d86-aa4d-020d8bbd5b83

**Comment by finnlewis** [comment]
I've tested this locally and it all seems to work OK. 

> When testing this branch with demo content I'm seeing this error
> 
> > The website encountered an unexpected error. Please try again later.
> > Error: Call to a member function getCacheTags() on null in Drupal\localgov_directories\Plugin\Block\ChannelSearchBlock->getCacheTags() (line 144 of modules/contrib/localgov_directories/src/Plugin/Block/ChannelSearchBlock.php)
> 
> When viewing citizen advice entry.

I tested with the demo content locally and could not replicate your error @andybroomfield

**Comment by finnlewis** [comment]
@j4-m is jumping in this week and will take a look at this pull request too. The more eyes the better, right?

**Comment by andybroomfield** [comment]
> I tested with the demo content locally and could not replicate your error @andybroomfield

@finnlewis full reinstall seems to solve it, maybe I have something weird in Lando database keeping things around on the test site?

**Comment by ekes** [comment]
So to update this. It's included in localgovdrupal/localgov#374 which with configuration updates is now passing.

**Comment by finnlewis** [comment]
Just to note, the tests are failing here as we don't have all the correct branches. 

The pull request here https://github.com/localgovdrupal/localgov/pull/375 shows all the tests passing.

**Review by finnlewis — APPROVED** [review]
Looks good to me. Safe enoiugh to try!

**Review by andybroomfield — APPROVED** [review]
I checked that installing this and the sub modules allowed directories to generate some mappings for open referral and I could see the json.
On local so wasn't able to validate it.

**Pull request merged**
